packages/marshal/src/marshal.js

```diff
@@ -69,9 +69,9 @@ export const makeMarshal = (
   };
 
   /**
-   * @type {Serialize<Slot>}
+   * @type {ToCapData<Slot>}
    */
-  const serialize = root => {
+  const toCapData = root => {
```

Comment:
This raises the interesting question of to what extent we consider "CapData" to describe just the `{ body: string, slots: any[] }` structure vs. that and/or a body string format that contains JSON text in which special values including those that cannot be directly represented in JSON are replaced with `{ [Symbol.for("@qclass")]: string, ... }` objects (as opposed to the "Smallcaps" body string format that inserts a "#" prefix before JSON text in which special values are represented as strings). Put another way, is `{ body: "{}", slots: [] }` something like CapData–CapData, or something like CapData–default/CapData–qclass?

Reply:
In choosing format names contrasting "capdata" vs "smallcaps", I feel like we've used up the term "capdata" for that narrower distinction. OTOH, we're still using the type name "CapData" for the record covering both these formats, and possibly future similar formats, when paired with a slots array. While I realize I caused this confusion, I think we need to repair it, and decide on one of these two meanings for the term "CapData".

I mildly favor resolving this by using "capdata" for the specific format and inventing another name for the record type. But I don't have a concrete suggestion. Also, we should expect the capdata format to be deprecated and eventually disappear, whereas the record type will continue, which perhaps argues for renaming the format rather than the type.

Reply:
> perhaps argues for renaming the format rather than the type.

+1 to this. If the format is to be deprecated in favor of "smallcaps" then we could name the format something that contrasts it with smallcaps. "oldcaps", "caps1", etc. "largecaps"! ;-)

`CapData` makes more sense to me as the generic record for whatever capability data, in whatever format it is encoded.

Reply:
API-wise, don't you want an API where the smallcap marshaller and the largecap marshaller both use the same API to produce their output?  In that case, this should be the more generic method name. I can see `toCapData` being generic enough.

Reply:
Now that @FUDCo 's smallcaps PR is merged, it may be easier to simply delete support for the deprecated format, quickly even if not immediately, rather than go through the pain of renaming it. It would be especially nice to delete it before bulldozer.

Reply:
But in any case, yeah I am convinced: Keep CapData as the name of the type and eat the pain on the format-name side.

Reply:
a `bodyFormat` property is compelling. Would you still call that structure `CapData`?

Reply:
If we consider `{ body: string, slots: any[] }` to be "CapData", then I think we would keep that name for `{ bodyFormat: string, body: any, slots: any[] }` (with a modifier if distinguishing old vs. new is important, e.g. "CapData2" or "CapData2023" or "flexible CapData" or "self-describing CapData").

Reply:
`bodyFormat: string` in every one of these records? Compactness matters here, which is why we did smallcaps in the first place, and why I chose to identify the difference in format with a single `'#'` character.

Reply:
There are several places where even the smallcaps encoding could be more compact (some of which have been explicitly considered and rejected, e.g. [non-decimal bigints](https://github.com/Agoric/agoric-sdk/pull/4469#issuecomment-1043602399)), but note that even in that dimension the fixed overhead of `"bodyFormat":"sigil",` would pay for itself in eliminated backslashes against any encoding that recursively includes at least 10 object properties or _x_ properties and _y_ string values where _x_ + _y_ ≥ 10:
```diff
diff current vs. proposed serialization of { string: "foo", promise, set: makeCopySet(["a", "b"]) }
-{"body":"{\"string\":\"foo\",\"promise\":\"&0\",\"set\":{\"#tag\":\"copySet\",\"payload\":[\"b\",\"a\"]}}","slots":["o+10"]}
+{"bodyFormat":"sigil","body":{"string":"foo","promise":"&0","set":{"#tag":"copySet","payload":["b","a"]}},"slots":["o+10"]}
```
(and even fewer than that with a smaller but more opaque alternative—`"f":1,` would break even at just two strings, and automatic inference when `typeof body !== 'string'` starts off two characters _smaller_ [at the cost of additional complexity to encode an unwrapped string which would otherwise be ambiguous with the current double serialization]).

Reply:
I have to admit, since we already separated the transformation from a passable into a serialization friendly structure, and the actual string serialization of that structure, it would make sense having a way to carry this structure around without forcing its serialization, allowing for more natural nesting.

However I'd like to point out that a smallcap structure nested inside another smallcap structure will still require nested marshalling, as any string key or value starting with a reserved char (ascii 33-45) will need to be `!` prefixed, so the nested gains are not as straightforward to calculate (but worst case scenario we'd have 1 extra char per string where with nested serialization we have 2).

We might also be able to avoid JSON serialization altogether as the final decision to serialize would be left to the transport layer. For example we could decide to adopt [tagged netstring](https://tnetstrings.info/) for the XS worker <-> kernel communication channel.

Reply:
The double encoding is indeed awful. We should figure out how to avoid the double encoding. But still, please let's not add a `bodyFormat` property.

Reply:
An alternative would be to use a different property name than `body` to hold structured / non serialized data. The property name would be representative of the encoding it contains.

For example `{ smallCapsBody: { ... }, slots: [] }`

Reply:
Anyone have reservations with landing this PR as is? https://github.com/Agoric/agoric-sdk/pull/6646 is sort of waiting on it.